pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
 // understand the consistency implications of having unpredictable numbers of pods available.
 func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, pods []*v1.Pod) error {
+
+	// list all revisions and sort them
+	revisions, err := ssc.ListRevisions(set)
+	if err != nil {
+		return err
+	}
+	history.SortControllerRevisions(revisions)
+
+	// get the current, and update revisions
+	currentRevision, updateRevision, err := ssc.getSetRevisions(set, revisions)
+	if err != nil {
+		return err
+	}
+
+	// perform the main update function and get the status
+	status, err := ssc.updateStatefulSet(set, currentRevision, updateRevision, pods)
+	if err != nil {
+		return err
+	}
+
+	// update the set's status
+	err = ssc.updateStatefulSetStatus(set, status)
+	if err != nil {
+		return err
+	}
+
+	// maintain the set's revision history limit
+	return ssc.truncateHistory(set, pods, revisions, currentRevision, updateRevision)
+}
+
+func (ssc *defaultStatefulSetControl) ListRevisions(set *apps.StatefulSet) ([]*apps.ControllerRevision, error) {
+	selector, err := metav1.LabelSelectorAsSelector(set.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	return ssc.controllerHistory.ListControllerRevisions(set, selector)
+}
+
+func (ssc *defaultStatefulSetControl) AdoptOrphanRevisions(
+	set *apps.StatefulSet,
+	revisions []*apps.ControllerRevision) error {
+	for i := range revisions {
+		adopted, err := ssc.controllerHistory.AdoptControllerRevision(set, controllerKind, revisions[i])
+		if err != nil {
+			return err
+		}
+		revisions[i] = adopted
+	}
+	return nil
+}
+
+// truncateHistory truncates any non-live ControllerRevisions in revisions from set's history. The UpdateRevision and
+// CurrentRevision in set's Status are considered to be live, as are any revisions associated with the Pods in pods.
+// Non-live revisions are deleted, starting with the revision with the lowest Revision, until only RevisionHistoryLimit
+// revisions remain. If the returned error is nil the operation was successful.
+func (ssc *defaultStatefulSetControl) truncateHistory(
+	set *apps.StatefulSet,
+	pods []*v1.Pod,
+	revisions []*apps.ControllerRevision,
+	current *apps.ControllerRevision,
+	update *apps.ControllerRevision) error {
+	history := make([]*apps.ControllerRevision, 0, len(revisions))
+	// mark all live revisions
+	live := map[string]bool{current.Name: true, update.Name: true}
+	for i := range pods {
+		live[getPodRevision(pods[i])] = true
+	}
+	// collect live revisions and historic revisions
+	for i := range revisions {
+		if !live[revisions[i].Name] {
+			history = append(history, revisions[i])
+		}
+	}
+	if len(history) < int(*set.Spec.RevisionHistoryLimit) {
+		return nil
+	}
+	// delete any non-live history to maintain the revision limit.
+	history = history[:int(*set.Spec.RevisionHistoryLimit)]
+	for i := 0; i < len(history); i++ {
+		if err := ssc.controllerHistory.DeleteControllerRevision(history[i]); err != nil {
+			return err
```

Comment:
No. There is a backoff throttle in the outer sync loop that causes retries to be delayed. That is why we bail out early on errors. If the API Server is stressed we shouldn't continue to pound it. We should back off and attempt again later.